cpp/tests/interop/nanoarrow_utils.hpp

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column_view.hpp>
+#include <cudf/interop/detail/arrow.hpp>
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/transform.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/traits.hpp>
+
+static ArrowBufferAllocator noop_alloc = (struct ArrowBufferAllocator){
+  .reallocate = [](ArrowBufferAllocator*, uint8_t* ptr, int64_t, int64_t) -> uint8_t* {
+    return ptr;
+  },
+  .free         = [](ArrowBufferAllocator*, uint8_t*, int64_t) {},
+  .private_data = nullptr,
+};
+
+template <typename T>
+std::enable_if_t<cudf::is_fixed_width<T>() and !std::is_same_v<T, bool>, void> get_nanoarrow_array(
+  ArrowArray* arr, std::vector<T> const& data, std::vector<uint8_t> const& mask = {})
+{
+  arr->length = data.size();
+  NANOARROW_THROW_NOT_OK(
+    ArrowBufferAppend(ArrowArrayBuffer(arr, 1), data.data(), sizeof(T) * data.size()));
+  if (!mask.empty()) {
+    NANOARROW_THROW_NOT_OK(ArrowBitmapReserve(ArrowArrayValidityBitmap(arr), mask.size()));
+    ArrowBitmapAppendInt8Unsafe(
+      ArrowArrayValidityBitmap(arr), reinterpret_cast<const int8_t*>(mask.data()), mask.size());
+    arr->null_count = ArrowBitCountSet(ArrowArrayValidityBitmap(arr)->buffer.data, 0, data.size());
+  } else {
+    arr->null_count = 0;
+  }
+
+  CUDF_EXPECTS(ArrowArrayFinishBuildingDefault(arr, nullptr) == NANOARROW_OK,
+               "failed to construct array");
+}
+
+template <typename T>
+std::enable_if_t<cudf::is_fixed_width<T>() and !std::is_same_v<T, bool>, void> populate_from_col(
+  ArrowArray* arr, cudf::column_view view)
```

Comment:
I think that would be more trouble than it's worth precisely because it would require us copying over the definition to make it useful. Not only that, but these functions are intended to be useful for children and recursive calling, if we expect it to be an `ArrowDeviceArray` we'd also have to pass the path down to the child we intend which could get extremely annoying . It's also much more convenient to have it just pass the `ArrowArray*` directly

Reply:
Sounds right, thanks.